### ISSUE TITLE
Sometimes ssgl_server fails to start

### ISSUE THREAD

**Issue by smcameron** [smcameron/space-nerds-in-space#201]
I noticed that on my desktop system, after I first boot it up, ssgl_server fails to start.  But subsequent attempts to start it work.  Don't know why.

**Comment by MCMic** [comment]
I can confirm the same problem. No log is produced.

Using snis_launcher I usually have to start the lobby twice.

**Comment by smcameron** [comment]
Managed to coax this out of it:

```
2019-02-23 10:54:39.863440: setsockopt() failed: Bad file descriptor
2019-02-23 10:54:39.863505: ssgl_server exiting, reason: setsockopt failed., exit code = 1
```

**Comment by smcameron** [comment]
And the reason for the bad file descriptor is that socket(2) failed (and why the hell wasn't I already checking the return value of socket()????  Bad programmer!  Bad!)

```
2019-02-23 11:02:41.728316: socket() failed: Protocol not supported
2019-02-23 11:02:41.728377: ssgl_server exiting, reason: socket() failed., exit code = 1
```

WTF?

The protocol is obtained from /etc/services via:

```
        /* Get the "gamelobby" service protocol/port */
        gamelobby_service = getservbyname(GAMELOBBY_SERVICE_NAME, "tcp");
        if (!gamelobby_service) {
                ssgl_log(SSGL_WARN, "getservbyname failed, %s\n", strerror(errno));
                ssgl_log(SSGL_WARN, "Check that /etc/services contains the following lines:\n");
                ssgl_log(SSGL_WARN, "gamelobby  2419/tcp\n");
                ssgl_log(SSGL_WARN, "gamelobby  2419/udp\n");
                ssgl_log(SSGL_WARN, "Continuing anyway, will assume port is 2419.\n");
        }

        /* Get the protocol number... */
        if (gamelobby_service) {
                gamelobby_proto = getprotobyname(gamelobby_service->s_proto);
                if (!gamelobby_proto) {
                        ssgl_log(SSGL_WARN, "getprotobyname(%s) failed: %s\n",
                                gamelobby_service->s_proto, strerror(errno));
                }
        } else {
                gamelobby_proto = NULL;
        }
```

and as such, it's not the kind of thing that should fail one time, then not fail on the 2nd try.  wtf?

**Comment by smcameron** [comment]
Curiouser and curiouser.

When it fails:

```
2019-02-23 11:23:31.961324: ssgl_server starting
2019-02-23 11:23:31.961422: ssgl_server: game server expiration thread started.
2019-02-23 11:23:31.975907: gamelobby_service = 0x7fe218d10480, gamelobby_proto = 0x7fe218d10440
2019-02-23 11:23:31.975915: protocol number is 17
2019-02-23 11:23:31.976059: comparing ipaddr 0100007f to 9d01a8c0
2019-02-23 11:23:31.976066: comparing ipaddr 9d01a8c0 to 9d01a8c0
2019-02-23 11:23:31.976068: Matched IP address 9d01a8c0
2019-02-23 11:23:31.976070: netmask is 00ffffff
2019-02-23 11:23:31.976072: broadcast is ff01a8c0
2019-02-23 11:23:31.977386: socket() failed: Protocol not supported
2019-02-23 11:23:31.977408: ssgl_server exiting, reason: socket() failed., exit code = 1
```

When it works:

```
2019-02-23 11:20:01.230413: ssgl_server starting
2019-02-23 11:20:01.230658: ssgl_server: game server expiration thread started.
2019-02-23 11:20:01.231247: gamelobby_service = 0x7f4fdcb07480, gamelobby_proto = 0x7f4fdcb07440
2019-02-23 11:20:01.231268: protocol number is 6
2019-02-23 11:20:01.231293: ssgl_server: listening for connections...
2019-02-23 11:20:01.231648: comparing ipaddr 0100007f to 9d01a8c0
2019-02-23 11:20:01.231674: comparing ipaddr 9d01a8c0 to 9d01a8c0
2019-02-23 11:20:01.231683: Matched IP address 9d01a8c0
2019-02-23 11:20:01.231691: netmask is 00ffffff
2019-02-23 11:20:01.231712: broadcast is ff01a8c0
```

6 and 17 are tcp and udp, respectively.

```
scameron@wombat ~ $ egrep 'tcp|udp' /etc/protocols
tcp	6	TCP		# transmission control protocol
udp	17	UDP		# user datagram protocol
udplite	136	UDPLite		# UDP-Lite [RFC3828]
scameron@wombat ~ $
```

It should always be getting tcp though, since we specifically asked getservbyname() for "tcp" ... , why is it getting udp on the first attempt after booting up?

**Comment by smcameron** [comment]
OOOOOOOOOH-K, I finally figured out what's going on.

getservbyname() and getprotobyname() are not thread safe! (Duh!)  They return pointers into static buffers. 

But I have two threads, the main thread, and broadcast_lobby_info.  One is using a udp socket, and the other a tcp socket.  They race, and sometimes the udp thread clobbers the tcp thread's return value from getservbyname("gamelobby", "tcp") with the return value from getservbyname("gamelobby", "udp");

So from the tcp thread's point of view, it asked for the "gamelobby/tcp" service but got the "gamelobby/udp" service.

**Comment by smcameron** [comment]
Fixed by 582671fd0beb11dfc5301e9ff805292c61694cad
